Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp

```diff
@@ -0,0 +1,1011 @@
+//       products derived from this software without specific prior written
+//       permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+#include "RGroupDecomp.h"
+
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/Substruct/SubstructMatch.h>
+#include <GraphMol/SmilesParse/SmilesWrite.h>
+#include <GraphMol/SmilesParse/SmartsWrite.h>
+#include <GraphMol/SmilesParse/SmilesParse.h>
+#include <GraphMol/ChemTransforms/ChemTransforms.h>
+#include <GraphMol/FMCS/FMCS.h>
+#include <set>
+#include <vector>
+
+//#define DEBUG
+
+namespace RDKit {
+
+// Attachment Points
+//  labeled cores => isotopes
+//  atom mappings
+//  atom indices => use -1 - atom index, range is [-1, ...., -num_atoms]
+namespace {
+const std::string RLABEL = "tempRlabel";
+const std::string SIDECHAIN_RLABELS = "sideChainRlabels";
+
+bool setLabel(Atom *atom, int label, std::set<int> &labels, int &maxLabel,
+              bool relabel, const std::string &type) {
+  if(type == "IsotopeLabels") {
+    atom->setIsotope(0);
+  }
+
+  if (label) {
+    if (labels.find(label) != labels.end()) {
+      if (relabel)
+        label = maxLabel + 1;
+      else
+        // XXX FIX me - get label id
+        throw ValueErrorException(
+            std::string("Duplicate label in input, current type is:") + type);
+    }
+
+    atom->setProp<int>(RLABEL, label);
+    labels.insert(label);
+    maxLabel = label + 1;
+    return true;
+  }
+  return false;
+}
+
+bool hasDummy(const RWMol &core) {
+  for (RWMol::ConstAtomIterator atIt = core.beginAtoms(); atIt != core.endAtoms();
+       atIt++) {
+    if((*atIt)->getAtomicNum() == 0) return true;
+  }
+  return false;
+}
+}
+
+bool RGroupDecompositionParameters::prepareCore(RWMol &core,
+                                                const RWMol *alignCore) {
+  const bool relabel = labels & RelabelDuplicateLabels;
+  if (alignCore && (alignment & MCS)) {
+    std::vector<ROMOL_SPTR> mols;
+    mols.push_back( ROMOL_SPTR(new ROMol(core)) );
+    mols.push_back( ROMOL_SPTR(new ROMol(*alignCore)) );
+    MCSResult res = findMCS(mols);
+    if(res.isCompleted()) {
+      const ROMol*m = (const ROMol*)SmartsToMol(res.SmartsString);
```

Comment:
Do I?  m is deleted below.  rdk_auto_ptr is better though, can't wait for unique_ptr myself :)